formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt

```diff
@@ -200,20 +247,98 @@ internal open class ProtobufDecoder(
 
     override fun SerialDescriptor.getTag(index: Int) = extractParameters(index)
 
+    private fun findUnreadElementIndex(): Int {
+        val elementsCount = descriptor.elementsCount
+        while (lowerReadMark != -1L) {
+            val index = lowerReadMark.inv().countTrailingZeroBits()
+            lowerReadMark = lowerReadMark or (1L shl index)
+
+            if (!descriptor.isElementOptional(index)) {
+                val elementDescriptor = descriptor.getElementDescriptor(index)
+                val kind = elementDescriptor.kind
+                if (kind == StructureKind.MAP || kind == StructureKind.LIST) {
+                    return index
+                } else if (elementDescriptor.isNullable) {
+                    valueIsNull = true
+                    return index
+                }
+            }
+        }
+
+        if (elementsCount > Long.SIZE_BITS) {
+            val higherMarks = highReadMarks!!
+
+            for (slot in higherMarks.indices) {
+                // (slot + 1) because first element in high marks has index 64
+                val slotOffset = (slot + 1) * Long.SIZE_BITS
+                // store in a variable so as not to frequently use the array
+                var mark = higherMarks[slot]
+
+                while (mark != -1L) {
```

Comment:
This loop can be extracted in a function as the previous one

Reply:
This loop differs in that it contains `higherMarks[slot] = mark` statement and I have not found a way to write simple and fast function that returns an indication that an absence value was found and mark value after loop.